r2-core/src/main/java/com/linkedin/r2/transport/http/client/EvictingCircularBuffer.java

```diff
@@ -0,0 +1,245 @@
+    thisLock.writeLock().lock();
+    try
+    {
+      _callbacks.set(writerPosition, toAdd);
+      _ttlBuffer.set(writerPosition, Instant.ofEpochMilli(_clock.currentTimeMillis()));
+    } finally {
+      thisLock.writeLock().unlock();
+    }
+  }
+
+  /**
+   * Returns a Callback previously stored in the circular buffer through the put() method. Callbacks are generally returned
+   * in the order they were received, but in cases of get() throughput in excess of put() throughput, previously returned
+   * Callbacks will be sent. Oldest age of returned Callbacks is configurable through the ttl param in the constructor.
+   *
+   * Calls to get() will always succeed as long as calls to put() continue at a cadence within the ttl duration. When write
+   * throughput has dropped to zero, get() will eventually throw NoSuchElementException once the circular buffer has become
+   * fully pruned through expired ttl.
+   * @return Callback
+   * @throws NoSuchElementException if internal circular buffer is empty
+   */
+  public Callback<None> get()
+  {
+    int thisReaderPosition = getAndBumpReaderPosition();
+    ReentrantReadWriteLock thisLock = _elementLocks.get(thisReaderPosition);
+
+    thisLock.readLock().lock();
+    Callback<None> callback;
+    Instant ttl;
+    try {
+      callback = _callbacks.get(thisReaderPosition);
+      ttl = _ttlBuffer.get(thisReaderPosition);
+    }
+    finally
+    {
+      thisLock.readLock().unlock();
+    }
+
+    if (callback == null)
+    {
+      if (thisReaderPosition == 0)
+      {
+        throw new NoSuchElementException("Buffer is empty");
+      }
+      else
+      {
+        return get();
+      }
+    }
+    else if (Duration.between(ttl, Instant.ofEpochMilli(_clock.currentTimeMillis())).compareTo(_ttl) > 0)
+    {
+      if (thisReaderPosition != 0 || this.size() == 1)
+      {
+        thisLock.writeLock().lock();
+        try {
+          if (callback == _callbacks.get(thisReaderPosition))
+          {
+            _callbacks.set(thisReaderPosition, null);
+            _ttlBuffer.set(thisReaderPosition, null);
+          }
+        }
+        finally
+        {
+          thisLock.writeLock().unlock();
+        }
+      }
+      return get();
+    }
+    return callback;
+  }
+
+  /**
+   * @return the number of unique Callbacks this buffer can hold.
+   */
+  int getCapacity()
+  {
+    return _callbacks.size();
+  }
+
+  /**
+   * Resizes the circular buffer, deleting the contents in the process.
+   * This method should not be called frequently, ideally only as part of a startup lifecycle, as it does heavy locking
+   * to ensure all reads and writes are drained coinciding with the resizing of the buffer.
+   * @param capacity
+   */
+  void setCapacity(int capacity)
+  {
+    if (capacity < 1)
+    {
+      throw new IllegalArgumentException("capacity can't be less than 1");
+    }
+    // acquire write lock for all elements in the buffer to prevent reads while the buffer is re-created,
```

Comment:
Spoke about this on Zoom. In summary, while we may not love these "change after instantiate" methods, they appear necessary in order to allow customers to change them through darkclusterconfig. If we wanted to get rid of them, the alternative would be passing the ratelimiter's components into the constructor of DarkClusterStrategyFactoryImpl instead of the ratelimiter itself, which is less clean than having these setters.